### PR TITLE
added spatial box/point, refactored contributor/creator

### PR DESCRIPTION
**Pull request by jo-pol**
@DANS-KNAW/easy 

Implemented conversion of SPATIAL columns in CSV file to DDM. Needed a slightly different approach than was implemented for contributors and creators and reafactored those into a similar new approach.

Support for DCX_RELATION columns have to follow in another pull request.

### CODE REVIEW

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddDatasetMetadataToDeposit.scala [hunk 1]
```diff
@@ -89,67 +85,44 @@ object AddDatasetMetadataToDeposit {
...
     val othersEmpty = authorFields
-      .filterNot(field => isOrganizationKey(field._1))
+      .filterNot(field => organizationKeys.contains(field._1))
       .forall(_._2 == "")
-    val hasOrganization = authorFields.toList.exists(field => isOrganizationKey(field._1))
+    val hasOrganization = authorFields.toList.exists(field => organizationKeys.contains(field._1))
     othersEmpty && hasOrganization
   }
 
-  def isOrganizationKey(key: MultiDepositKey) = key match {
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddDatasetMetadataToDeposit.scala:135]: Good catch of hardcoding refactoring 👍

File: src/main/scala/nl/knaw/dans/easy/multideposit/package.scala [hunk 2]
```diff
@@ -255,6 +255,20 @@ package object multideposit {
         value2 <- value.toOption
       } yield value2
     }
+
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/multideposit/package.scala:259]: Some more documentation on these 4 methods wouldn't hurt IMHO. As both `Dataset` and `Dictionary` are somewhat more complex data structures, this documentation would get the functionality of these methods across a little faster to the others.

**Comment by rvanheest** [comment]
While running it, I encountered a `NullPointerException`...

`./run.sh -s target/springfield target/test-classes/Roundtrip_MD/sip-demo-2015-02-24 target/result`

gave the following output:

```
./run.sh: line 21: apphome.sh: No such file or directory
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building EASY Split Multi-Deposit 1.x-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- exec-maven-plugin:1.4.0:java (default-cli) @ easy-split-multi-deposit ---
[INFO ] Compiling list of actions to perform ...
[INFO ] Checking preconditions of CreateOutputDepositDir ...
[INFO ] Checking preconditions of AddBagToDeposit ...
[INFO ] Checking preconditions of AddDatasetMetadataToDeposit ...
[INFO ] Checking preconditions of AddFileMetadataToDeposit ...
[INFO ] Checking preconditions of AddPropertiesToDeposit ...
[INFO ] Checking preconditions of CopyToSpringfieldInbox ...
[INFO ] Checking preconditions of CreateOutputDepositDir ...
[INFO ] Checking preconditions of AddBagToDeposit ...
[INFO ] Checking preconditions of AddDatasetMetadataToDeposit ...
[INFO ] Checking preconditions of AddFileMetadataToDeposit ...
[INFO ] Checking preconditions of AddPropertiesToDeposit ...
[INFO ] Checking preconditions of CopyToSpringfieldInbox ...
[INFO ] Checking preconditions of CreateSpringfieldActions ...
[INFO ] Executing action of CreateOutputDepositDir ...
[INFO ] Executing action of AddBagToDeposit ...
[INFO ] Executing action of AddDatasetMetadataToDeposit ...
java.lang.NullPointerException
    at scala.collection.mutable.HashTable$$anon$1.next(HashTable.scala:214)
    at scala.collection.mutable.HashTable$$anon$1.next(HashTable.scala:206)
    at scala.collection.mutable.HashMap$$anon$4.next(HashMap.scala:122)
    at scala.collection.IterableLike$class.head(IterableLike.scala:107)
    at scala.collection.AbstractIterable.head(Iterable.scala:54)
    at nl.knaw.dans.easy.multideposit.package$DatasetExtensions$.toRows$extension(package.scala:269)
    at nl.knaw.dans.easy.multideposit.package$DatasetExtensions$.rowsWithValuesFor$extension(package.scala:260)
    at nl.knaw.dans.easy.multideposit.actions.AddDatasetMetadataToDeposit$.createContributors(AddDatasetMetadataToDeposit.scala:114)
    at nl.knaw.dans.easy.multideposit.actions.AddDatasetMetadataToDeposit$.createMetadata(AddDatasetMetadataToDeposit.scala:170)
    at nl.knaw.dans.easy.multideposit.actions.AddDatasetMetadataToDeposit$.datasetToXml(AddDatasetMetadataToDeposit.scala:64)
    at nl.knaw.dans.easy.multideposit.actions.AddDatasetMetadataToDeposit$$anonfun$writeDatasetMetadataXml$1.apply$mcV$sp(AddDatasetMetadataToDeposit.scala:43)
    at nl.knaw.dans.easy.multideposit.actions.AddDatasetMetadataToDeposit$$anonfun$writeDatasetMetadataXml$1.apply(AddDatasetMetadataToDeposit.scala:43)
    at nl.knaw.dans.easy.multideposit.actions.AddDatasetMetadataToDeposit$$anonfun$writeDatasetMetadataXml$1.apply(AddDatasetMetadataToDeposit.scala:43)
    at scala.util.Try$.apply(Try.scala:192)
    at nl.knaw.dans.easy.multideposit.actions.AddDatasetMetadataToDeposit$.writeDatasetMetadataXml(AddDatasetMetadataToDeposit.scala:42)
    at nl.knaw.dans.easy.multideposit.actions.AddDatasetMetadataToDeposit.run(AddDatasetMetadataToDeposit.scala:37)
    at nl.knaw.dans.easy.multideposit.Main$$anonfun$runActions$2.apply(Main.scala:186)
    at nl.knaw.dans.easy.multideposit.Main$$anonfun$runActions$2.apply(Main.scala:186)
    at rx.lang.scala.Observable$$anon$19.call(Observable.scala:1090)
    at rx.lang.scala.Observable$$anon$19.call(Observable.scala:1089)
    at rx.internal.operators.OperatorMap$1.onNext(OperatorMap.java:54)
    at rx.internal.operators.OperatorDoOnEach$1.onNext(OperatorDoOnEach.java:85)
    at rx.internal.operators.OperatorMerge$MergeSubscriber.emitScalar(OperatorMerge.java:365)
    at rx.internal.operators.OperatorMerge$MergeSubscriber.tryEmit(OperatorMerge.java:327)
    at rx.internal.operators.OperatorMerge$InnerSubscriber.onNext(OperatorMerge.java:804)
    at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.slowpath(OnSubscribeFromIterable.java:97)
    at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.request(OnSubscribeFromIterable.java:73)
    at rx.Subscriber.setProducer(Subscriber.java:211)
    at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:49)
    at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:32)
    at rx.Observable.unsafeSubscribe(Observable.java:8098)
    at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:232)
    at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:142)
    at rx.internal.operators.OperatorMap$1.onNext(OperatorMap.java:54)
    at rx.internal.operators.OperatorTakeLastOne$ParentSubscriber.emit(OperatorTakeLastOne.java:152)
    at rx.internal.operators.OperatorTakeLastOne$ParentSubscriber.onCompleted(OperatorTakeLastOne.java:125)
    at rx.internal.operators.OperatorScan$InitialProducer.checkTerminated(OperatorScan.java:218)
    at rx.internal.operators.OperatorScan$InitialProducer.emitLoop(OperatorScan.java:307)
    at rx.internal.operators.OperatorScan$InitialProducer.emit(OperatorScan.java:293)
    at rx.internal.operators.OperatorScan$InitialProducer.onCompleted(OperatorScan.java:235)
    at rx.internal.operators.OperatorScan$3.onCompleted(OperatorScan.java:156)
    at rx.internal.operators.OperatorDoOnEach$1.onCompleted(OperatorDoOnEach.java:53)
    at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:614)
    at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:526)
    at rx.internal.operators.OperatorMerge$MergeSubscriber.onCompleted(OperatorMerge.java:255)
    at rx.internal.operators.OperatorMap$1.onCompleted(OperatorMap.java:43)
    at rx.internal.operators.OperatorTakeLastOne$ParentSubscriber.emit(OperatorTakeLastOne.java:159)
    at rx.internal.operators.OperatorTakeLastOne$ParentSubscriber.onCompleted(OperatorTakeLastOne.java:125)
    at rx.internal.operators.OperatorScan$InitialProducer.checkTerminated(OperatorScan.java:218)
    at rx.internal.operators.OperatorScan$InitialProducer.emitLoop(OperatorScan.java:307)
    at rx.internal.operators.OperatorScan$InitialProducer.emit(OperatorScan.java:293)
    at rx.internal.operators.OperatorScan$InitialProducer.onCompleted(OperatorScan.java:235)
    at rx.internal.operators.OperatorScan$3.onCompleted(OperatorScan.java:156)
    at rx.internal.operators.OperatorMap$1.onCompleted(OperatorMap.java:43)
    at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:614)
    at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:526)
    at rx.internal.operators.OperatorMerge$MergeSubscriber.onCompleted(OperatorMerge.java:255)
    at rx.internal.operators.OperatorMap$1.onCompleted(OperatorMap.java:43)
    at rx.observers.SafeSubscriber.onCompleted(SafeSubscriber.java:84)
    at rx.internal.operators.OperatorMap$1.onCompleted(OperatorMap.java:43)
    at rx.internal.operators.OperatorTakeLastOne$ParentSubscriber.emit(OperatorTakeLastOne.java:159)
    at rx.internal.operators.OperatorTakeLastOne$ParentSubscriber.onCompleted(OperatorTakeLastOne.java:125)
    at rx.internal.operators.OperatorScan$InitialProducer.checkTerminated(OperatorScan.java:218)
    at rx.internal.operators.OperatorScan$InitialProducer.emitLoop(OperatorScan.java:307)
    at rx.internal.operators.OperatorScan$InitialProducer.emit(OperatorScan.java:293)
    at rx.internal.operators.OperatorScan$InitialProducer.onCompleted(OperatorScan.java:235)
    at rx.internal.operators.OperatorScan$3.onCompleted(OperatorScan.java:156)
    at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.slowpath(OnSubscribeFromIterable.java:101)
    at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.request(OnSubscribeFromIterable.java:73)
    at rx.internal.operators.OperatorScan$InitialProducer.setProducer(OperatorScan.java:280)
    at rx.internal.operators.OperatorScan$3.setProducer(OperatorScan.java:161)
    at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:49)
    at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:32)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable.subscribe(Observable.java:8191)
    at rx.Observable.subscribe(Observable.java:8158)
    at rx.lang.scala.Observable$class.subscribe(Observable.scala:178)
    at rx.lang.scala.JavaConversions$$anon$2.subscribe(JavaConversions.scala:53)
    at nl.knaw.dans.easy.multideposit.MultiDepositParser$$anonfun$parse$1$$anonfun$apply$1.apply(MultiDepositParser.scala:62)
    at nl.knaw.dans.easy.multideposit.MultiDepositParser$$anonfun$parse$1$$anonfun$apply$1.apply(MultiDepositParser.scala:43)
    at scala.util.Try$.apply(Try.scala:192)
    at nl.knaw.dans.easy.multideposit.MultiDepositParser$$anonfun$parse$1.apply(MultiDepositParser.scala:43)
    at nl.knaw.dans.easy.multideposit.MultiDepositParser$$anonfun$parse$1.apply(MultiDepositParser.scala:41)
    at rx.lang.scala.ImplicitFunctionConversions$$anon$1.call(ImplicitFunctionConversions.scala:53)
    at rx.lang.scala.ImplicitFunctionConversions$$anon$1.call(ImplicitFunctionConversions.scala:51)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable$2.call(Observable.java:162)
    at rx.Observable$2.call(Observable.java:154)
    at rx.Observable.subscribe(Observable.java:8191)
    at rx.Observable.subscribe(Observable.java:8158)
    at rx.Observable.subscribe(Observable.java:7872)
    at rx.lang.scala.Observable$class.subscribe(Observable.scala:139)
    at rx.lang.scala.JavaConversions$$anon$2.subscribe(JavaConversions.scala:53)
    at nl.knaw.dans.easy.multideposit.Main$.main(Main.scala:39)
    at nl.knaw.dans.easy.multideposit.Main.main(Main.scala)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:497)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:293)
    at java.lang.Thread.run(Thread.java:745)
[ERROR] An error occurred. Rolling back actions ...
[ERROR] Could not write dataset metadata: java.lang.NullPointerException
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 2.517 s
[INFO] Finished at: 2016-05-25T09:32:58+02:00
[INFO] Final Memory: 22M/497M
[INFO] ------------------------------------------------------------------------
```

**Comment by rvanheest** [comment]
Here is the test that detects the `NullPointerException`. @jo-pol, you can put this in `AddDatasetMetadataToDepositSpec`, where `it` refers to `datasetToXml`.

``` scala
it should "return valid xml on reading from the csv" in {
  val csv = new File(getClass.getResource("/Roundtrip_MD/sip-demo-2015-02-24/instructions.csv").toURI)
  val subscriber = TestSubscriber[Elem]
  MultiDepositParser.parse(csv)
    .flatMap(_.toObservable)
    .map(tuple => AddDatasetMetadataToDeposit.datasetToXml(tuple._2))
    .subscribe(subscriber)

//  subscriber.assertValueCount(2)
  subscriber.assertNoErrors()
  subscriber.assertCompleted()
  subscriber.assertUnsubscribed()
}
```

**Comment by rvanheest** [comment]
Here is a solution:

``` scala
def toRows = dataset.values.headOption
  .map(_.indices
    .map(i => dataset.map { case (key, values) => (key, values(i)) })
    .map(_.filter(kv => kv._2 != null && !kv._2.isBlank)))
  .getOrElse(IndexedSeq())
```

**Comment by jo-pol** [comment]
@janvanmansum @lindareijnhoudt 

**DON'T SQUASH-merge**, otherwise the two steps to get rid of the capital in the package folder would nullify one-another.

**Comment by rvanheest** [comment]
With the newest changes the `NullPointerException` is fixed. I approve of this PR. Great work @jo-pol!

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddDatasetMetadataToDeposit.scala [hunk 3]
```diff
@@ -160,13 +160,25 @@ object AddDatasetMetadataToDeposit {
     )
   }
 
+  def createRelations(dataset: Dataset) = {
+    dataset.rowsWithValuesFor(composedRelationFields).map { row =>
+      (row.get("DCX_RELATION_QUALIFIER"), row.get("DCX_RELATION_LINK"), row.get("DCX_RELATION_TITLE")) match {
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddDatasetMetadataToDeposit.scala:165]: Deze match is niet volledig: `(Some, None, None)` mist.

File: src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddDatasetMetadataToDeposit.scala [hunk 4]
```diff
@@ -160,13 +160,25 @@ object AddDatasetMetadataToDeposit {
     )
   }
 
+  def createRelations(dataset: Dataset) = {
+    dataset.rowsWithValuesFor(composedRelationFields).map { row =>
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddDatasetMetadataToDeposit.scala:164]: `rowsWithValuesFor` is een relatief dure operatie. Kan deze methode niet op een andere manier worden geïmplementeerd?

Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/multideposit/actions/AddDatasetMetadataToDeposit.scala:164]: ik ga hier even naar kijken ;-)

**Pull request merged**